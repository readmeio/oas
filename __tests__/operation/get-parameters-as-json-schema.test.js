/* eslint-disable jsdoc/require-jsdoc */
const Oas = require('../../src').default;

const createOas = require('../__fixtures__/create-oas').default;
const circular = require('../__datasets__/circular.json');
const discriminators = require('../__datasets__/discriminators.json');
const petstore = require('@readme/oas-examples/3.0/json/petstore.json');
const petstoreServerVars = require('../__datasets__/petstore-server-vars.json');
const deprecated = require('../__datasets__/schema-deprecated.json');

test('it should return with null if there are no parameters', () => {
  expect(createOas({ parameters: [] }).operation('/', 'get').getParametersAsJsonSchema()).toBeNull();
  expect(createOas({}).operation('/', 'get').getParametersAsJsonSchema()).toBeNull();
});

describe('type sorting', () => {
  const operation = {
    parameters: [
      { in: 'path', name: 'path parameter', schema: { type: 'string' } },
      { in: 'query', name: 'query parameter', schema: { type: 'string' } },
      { in: 'header', name: 'header parameter', schema: { type: 'string' } },
      { in: 'cookie', name: 'cookie parameter', schema: { type: 'string' } },
    ],
    requestBody: {
      description: 'Body description',
      content: {},
    },
  };

  it('should return with a json schema for each parameter type (formData instead of body)', () => {
    operation.requestBody.content = {
      'application/x-www-form-urlencoded': {
        schema: {
          type: 'object',
          properties: { a: { type: 'string' } },
        },
      },
    };

    const oas = createOas(operation);
    const jsonschema = oas.operation('/', 'get').getParametersAsJsonSchema();

    expect(jsonschema).toMatchSnapshot();
    expect(
      jsonschema.map(js => {
        return js.type;
      })
    ).toStrictEqual(['path', 'query', 'cookie', 'formData', 'header']);
  });

  it('should return with a json schema for each parameter type (body instead of formData)', () => {
    operation.requestBody.content = {
      'application/json': {
        schema: {
          type: 'object',
          properties: { a: { type: 'string' } },
        },
      },
    };

    const oas = createOas(operation);
    const jsonschema = oas.operation('/', 'get').getParametersAsJsonSchema();

    expect(jsonschema).toMatchSnapshot();
    expect(
      jsonschema.map(js => {
        return js.type;
      })
    ).toStrictEqual(['path', 'query', 'body', 'cookie', 'header']);
  });
});

describe('parameters', () => {
  it('should convert parameters to JSON schema', async () => {
    const oas = new Oas(petstore);
    await oas.dereference();

    const operation = oas.operation('/pet/{petId}', 'delete');
    expect(operation.getParametersAsJsonSchema()).toMatchSnapshot();
  });

  describe('type quirks', () => {
    it('should set a type to `string` if neither `schema` or `current` are present', () => {
      const oas = createOas({
        parameters: [
          {
            name: 'userId',
            in: 'query',
          },
        ],
      });

      const schema = oas.operation('/', 'get').getParametersAsJsonSchema();
      expect(schema[0].schema.properties.userId).toStrictEqual({
        type: 'string',
      });
    });
  });

  describe('`content` support', () => {
    it('should support `content` on parameters', () => {
      const oas = createOas({
        parameters: [
          {
            name: 'userId',
            description: 'Filter the data by userId',
            in: 'query',
            content: { 'application/json': { schema: { type: 'string' } } },
          },
        ],
      });

      const schema = oas.operation('/', 'get').getParametersAsJsonSchema();
      expect(schema[0].schema.properties.userId).toStrictEqual({
        $schema: 'http://json-schema.org/draft-04/schema#',
        type: 'string',
        description: 'Filter the data by userId',
      });
    });

    it('should prioritize `application/json` if present', () => {
      const oas = createOas({
        parameters: [
          {
            name: 'userId',
            in: 'query',
            content: {
              'application/json': { schema: { type: 'integer' } },
              'application/xml': { schema: { type: 'string' } },
            },
          },
        ],
      });

      const schema = oas.operation('/', 'get').getParametersAsJsonSchema();
      expect(schema[0].schema.properties.userId).toStrictEqual({
        $schema: 'http://json-schema.org/draft-04/schema#',
        type: 'integer',
      });
    });

    it("should prioritize JSON-like content types if they're present", () => {
      const oas = createOas({
        parameters: [
          {
            name: 'userId',
            in: 'query',
            content: {
              // Though is the first entry here is XML, we should actually use the second instead because it's
              // JSON-like.
              'application/xml': { schema: { type: 'string' } },
              'application/vnd.github.v3.star+json': {
                schema: { type: 'integer' },
              },
            },
          },
        ],
      });

      const schema = oas.operation('/', 'get').getParametersAsJsonSchema();
      expect(schema[0].schema.properties.userId).toStrictEqual({
        $schema: 'http://json-schema.org/draft-04/schema#',
        type: 'integer',
      });
    });

    it('should use the first content type if `application/json` is not present', () => {
      const oas = createOas({
        parameters: [
          {
            name: 'userId',
            in: 'query',
            content: {
              'application/xml': { schema: { type: 'integer' } },
              'text/plain': { schema: { type: 'string' } },
            },
          },
        ],
      });

      const schema = oas.operation('/', 'get').getParametersAsJsonSchema();
      expect(schema[0].schema.properties.userId).toStrictEqual({
        $schema: 'http://json-schema.org/draft-04/schema#',
        type: 'integer',
      });
    });
  });

  describe('common parameters', () => {
    it('should override path-level parameters on the operation level', () => {
      const oas = new Oas({
        paths: {
          '/pet/{petId}': {
            parameters: [
              {
                name: 'petId',
                in: 'path',
                description: 'ID of pet to return',
                schema: {
                  type: 'string',
                },
                required: true,
              },
            ],
            get: {
              parameters: [
                {
                  name: 'petId',
                  in: 'path',
                  description: 'A comma-separated list of pet IDs',
                  schema: {
                    type: 'string',
                  },
                  required: true,
                },
              ],
            },
          },
        },
      });

      expect(
        oas.operation('/pet/{petId}', 'get').getParametersAsJsonSchema()[0].schema.properties.petId.description
      ).toBe('A comma-separated list of pet IDs');
    });

    it('should add common parameter to path params', () => {
      const oas = new Oas({
        paths: {
          '/pet/{petId}': {
            parameters: [
              {
                name: 'petId',
                in: 'path',
                description: 'ID of pet to return',
                schema: {
                  type: 'string',
                },
                required: true,
              },
            ],
            get: {},
          },
        },
      });

      const operation = oas.operation('/pet/{petId}', 'get');

      expect(operation.getParametersAsJsonSchema()[0].schema.properties.petId.description).toBe(
        oas.api.paths['/pet/{petId}'].parameters[0].description
      );
    });

    it('should handle a common parameter dereferenced $ref to path params', async () => {
      const oas = new Oas({
        paths: {
          '/pet/{petId}': {
            parameters: [
              {
                $ref: '#/components/parameters/petId',
              },
            ],
            get: {},
          },
        },
        components: {
          parameters: {
            petId: {
              name: 'petId',
              in: 'path',
              description: 'ID of pet to return',
              schema: {
                type: 'string',
              },
              required: true,
            },
          },
        },
      });

      await oas.dereference();

      expect(
        oas.operation('/pet/{petId}', 'get').getParametersAsJsonSchema()[0].schema.properties.petId.description
      ).toBe(oas.api.components.parameters.petId.description);
    });
  });
});

describe('request bodies', () => {
  it('should convert request bodies to JSON schema (application/json)', async () => {
    const oas = new Oas(petstore);
    await oas.dereference();

    const operation = oas.operation('/pet', 'post');
    expect(operation.getParametersAsJsonSchema()).toMatchSnapshot();
  });

  it('should convert request bodies to JSON schema (application/x-www-form-urlencoded)', async () => {
    const oas = new Oas(petstoreServerVars);
    await oas.dereference();

    const operation = oas.operation('/pet/{petId}', 'post');
    expect(operation.getParametersAsJsonSchema()).toMatchSnapshot();
  });

  it('should not return anything for an empty schema', () => {
    const oas = createOas({
      requestBody: {
        description: 'Body description',
        content: {
          'application/json': {
            schema: {},
          },
        },
      },
    });

    expect(oas.operation('/', 'get').getParametersAsJsonSchema()).toStrictEqual([]);
  });
});

describe('$ref quirks', () => {
  it("should retain $ref pointers in the schema even if they're circular", async () => {
    const oas = new Oas(circular);
    await oas.dereference();

    expect(oas.operation('/', 'put').getParametersAsJsonSchema()).toMatchSnapshot();
  });
});

describe('polymorphism / discriminators', () => {
  it('should retain discriminator `mapping` refs when present', async () => {
    const oas = new Oas(discriminators);
    await oas.dereference();

    const operation = oas.operation('/anything/discriminator-with-mapping', 'patch');
    expect(operation.getParametersAsJsonSchema()).toMatchSnapshot();
  });
});

describe('type', () => {
  describe('request bodies', () => {
    describe('repair invalid schema that has no `type`', () => {
      it('should not add a string type on a `requestBody` and component schema that are clearly objects', () => {
        const oas = createOas(
          {
            requestBody: {
              content: {
                'application/json': {
                  schema: {
                    properties: {
                      uri: {
                        type: 'string',
                        format: 'uri',
                      },
                      messages: {
                        type: 'array',
                        items: {
                          $ref: '#/components/schemas/messages',
                        },
                      },
                      user: {
                        $ref: '#/components/schemas/user',
                      },
                    },
                  },
                },
              },
            },
          },
          {
            schemas: {
              messages: {
                properties: {
                  message: {
                    type: 'string',
                  },
                },
              },
              user: {
                required: ['user_id'],
                properties: {
                  user_id: {
                    type: 'integer',
                  },
                },
              },
            },
          }
        );

        // So we can test that components are transformed, this test intentionally does **not** dereference the API
        // definition.
        const schema = oas.operation('/', 'get').getParametersAsJsonSchema();

        expect(schema[0].schema.components.schemas.messages.type).toBe('object');
        expect(schema[0].schema.components.schemas.user.type).toBe('object');
      });
    });
  });
});

describe('descriptions', () => {
  it.todo('should pass through description on requestBody');

  it('should pass through description on parameters', () => {
    const oas = createOas({
      parameters: [
        {
          in: 'header',
          name: 'Accept',
          description: 'Expected response format.',
          schema: {
            type: 'string',
          },
        },
      ],
    });

    expect(oas.operation('/', 'get').getParametersAsJsonSchema()).toStrictEqual([
      {
        label: 'Headers',
        type: 'header',
        schema: {
          type: 'object',
          properties: {
            Accept: {
              $schema: 'http://json-schema.org/draft-04/schema#',
              description: 'Expected response format.',
              type: 'string',
            },
          },
          required: [],
        },
      },
    ]);
  });

  it('should pass through description on parameter when referenced as a `$ref` and a `requestBody` is present', async () => {
    const oas = createOas(
      {
        parameters: [
          {
            $ref: '#/components/parameters/pathId',
          },
        ],
        requestBody: {
          content: {
            'application/json': {
              schema: {
                type: 'object',
              },
            },
          },
        },
      },
      {
        parameters: {
          pathId: {
            name: 'pathId',
            in: 'path',
            description: 'Description for the pathId',
            required: true,
            schema: {
              type: 'integer',
              format: 'uint32',
            },
          },
        },
      }
    );

    await oas.dereference();

    expect(oas.operation('/', 'get').getParametersAsJsonSchema()[0].schema).toStrictEqual({
      type: 'object',
      properties: {
        pathId: {
          $schema: 'http://json-schema.org/draft-04/schema#',
          type: 'integer',
          format: 'uint32',
          maximum: 4294967295,
          minimum: 0,
          description: 'Description for the pathId',
        },
      },
      required: ['pathId'],
    });
  });
});

describe('required', () => {
  it.todo('should pass through `required` on parameters');

  it.todo('should make things required correctly for request bodies');
});

describe('`globalDefaults` option', () => {
  it('should use user defined `globalDefaults` for requestBody', async () => {
    const schema = new Oas(petstore);
    await schema.dereference();
    const operation = schema.operation('/pet', 'post');
    const jwtDefaults = {
      category: {
        id: 4,
        name: 'Testing',
      },
    };

    const jsonSchema = operation.getParametersAsJsonSchema(jwtDefaults);
    expect(jsonSchema[0].schema.properties.category.default).toStrictEqual(jwtDefaults.category);
  });

  it('should use user defined `globalDefaults` for parameters', async () => {
    const schema = new Oas(petstore);
    await schema.dereference();
    const operation = schema.operation('/pet/{petId}', 'get');
    const jwtDefaults = {
      petId: 1,
    };

    const jsonSchema = operation.getParametersAsJsonSchema(jwtDefaults);
    expect(jsonSchema[0].schema.properties.petId.default).toStrictEqual(jwtDefaults.petId);
  });
});

describe('`example` / `examples` support', () => {
  describe('parameters', () => {
    it.each([['example'], ['examples']])('should pick up %s if declared outside of the schema', exampleProp => {
      function createExample(value, inSchema = false) {
        if (exampleProp === 'example') {
          return value;
        }

        if (inSchema) {
          return [value];
        }

        return {
          distinctName: {
            value,
          },
        };
      }

      const oas = createOas({
        parameters: [
          {
            in: 'query',
            name: 'query parameter',
            schema: {
              type: 'string',
            },
            [exampleProp]: createExample('example foo'),
          },
          {
            in: 'query',
            name: 'query parameter alt',
            schema: {
              type: 'string',
              [exampleProp]: createExample('example bar', true),
            },
          },
        ],
      });

      const schema = oas.operation('/', 'get').getParametersAsJsonSchema();
      expect(schema).toStrictEqual([
        {
          type: 'query',
          label: 'Query Params',
          schema: {
            type: 'object',
            properties: {
              'query parameter': {
                $schema: 'http://json-schema.org/draft-04/schema#',
                type: 'string',
                examples: ['example foo'],
              },
              'query parameter alt': {
                $schema: 'http://json-schema.org/draft-04/schema#',
                type: 'string',
                examples: ['example bar'],
              },
            },
            required: [],
          },
        },
      ]);
    });
  });
});

describe('deprecated', () => {
  describe('parameters', () => {
    it('should pass through deprecated on parameters', () => {
      const oas = createOas({
        parameters: [
          {
            in: 'header',
            name: 'Accept',
            deprecated: true,
            schema: {
              type: 'string',
            },
          },
        ],
      });

      expect(oas.operation('/', 'get').getParametersAsJsonSchema()).toStrictEqual([
        {
          label: 'Headers',
          type: 'header',
          schema: {
            type: 'object',
            properties: {},
            required: [],
          },
          deprecatedProps: {
            type: 'header',
            schema: {
              type: 'object',
              $schema: 'http://json-schema.org/draft-04/schema#',
              properties: {
                Accept: {
                  $schema: 'http://json-schema.org/draft-04/schema#',
                  type: 'string',
                  deprecated: true,
                },
              },
              required: [],
            },
          },
        },
      ]);
    });

    it('should pass through deprecated on parameter when referenced as a `$ref` and a `requestBody` is present', async () => {
      const oas = createOas(
        {
          parameters: [
            {
              $ref: '#/components/parameters/pathId',
            },
          ],
          requestBody: {
            content: {
              'application/json': {
                schema: {
                  type: 'object',
                },
              },
            },
          },
        },
        {
          parameters: {
            pathId: {
              name: 'pathId',
              in: 'path',
              deprecated: true,
              schema: {
                type: 'integer',
                format: 'uint32',
              },
            },
          },
        }
      );

      await oas.dereference();
      expect(oas.operation('/', 'get').getParametersAsJsonSchema()[0].deprecatedProps.schema).toStrictEqual({
        type: 'object',
        $schema: 'http://json-schema.org/draft-04/schema#',
        properties: {
          pathId: {
            $schema: 'http://json-schema.org/draft-04/schema#',
            type: 'integer',
            format: 'uint32',
            maximum: 4294967295,
            minimum: 0,
            deprecated: true,
          },
        },
        required: [],
      });
    });

    it('should create deprecatedProps from body and metadata parameters', async () => {
      const oas = new Oas(deprecated);
      await oas.dereference();
      const operation = oas.operation('/anything', 'post');

      expect(operation.getParametersAsJsonSchema()).toMatchSnapshot();
    });

    it('should not put required deprecated parameters in deprecatedProps', async () => {
      const oas = new Oas(deprecated);
      await oas.dereference();
      const operation = oas.operation('/anything', 'post');
      const deprecatedSchema = operation.getParametersAsJsonSchema()[1].deprecatedProps.schema;

      deprecatedSchema.required.forEach(requiredParam => {
        expect(requiredParam in deprecatedSchema.properties).toBe(false);
      });
      expect(Object.keys(deprecatedSchema.properties)).toHaveLength(4);
    });

    it('should not put readOnly deprecated parameters in deprecatedProps', async () => {
      const oas = new Oas(deprecated);
      await oas.dereference();
      const operation = oas.operation('/anything', 'post');
      const deprecatedSchema = operation.getParametersAsJsonSchema()[1].deprecatedProps.schema;

      expect(Object.keys(deprecatedSchema.properties)).toHaveLength(4);
      expect('idReadOnly' in Object.keys(deprecatedSchema.properties)).toBe(false);
    });
  });

  describe('request bodies', () => {
    it('should pass through deprecated on a request body schema property', () => {
      const oas = createOas({
        requestBody: {
          content: {
            'application/json': {
              schema: {
                properties: {
                  uri: {
                    type: 'string',
                    format: 'uri',
                  },
                  messages: {
                    type: 'array',
                    items: {
                      type: 'string',
                    },
                    deprecated: true,
                  },
                },
              },
            },
          },
        },
      });

      expect(oas.operation('/', 'get').getParametersAsJsonSchema()).toStrictEqual([
        {
          type: 'body',
          label: 'Body Params',
          schema: {
            $schema: 'http://json-schema.org/draft-04/schema#',
            properties: {
              uri: { type: 'string', format: 'uri' },
            },
            type: 'object',
          },
          deprecatedProps: {
            type: 'body',
            schema: {
              $schema: 'http://json-schema.org/draft-04/schema#',
              properties: {
                messages: {
                  type: 'array',
                  items: {
                    type: 'string',
                  },
                  deprecated: true,
                },
              },
              type: 'object',
            },
          },
        },
      ]);
    });
  });

  describe('polymorphism', () => {
    it('should pass through deprecated on an allOf schema', () => {
      const oas = createOas({
        requestBody: {
          content: {
            'application/json': {
              schema: {
                allOf: [
                  {
                    type: 'object',
                    properties: {
                      uri: {
                        type: 'string',
                        format: 'uri',
                      },
                    },
                  },
                  {
                    type: 'object',
                    properties: {
                      messages: {
                        type: 'array',
                        items: {
                          type: 'string',
                        },
                        deprecated: true,
                      },
                    },
                  },
                ],
              },
            },
          },
        },
      });

      expect(oas.operation('/', 'get').getParametersAsJsonSchema()).toMatchSnapshot();
    });
  });
});
