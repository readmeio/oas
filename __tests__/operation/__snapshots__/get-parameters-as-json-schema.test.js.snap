// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`$ref quirks should retain $ref pointers in the schema even if they're circular 1`] = `
Array [
  Object {
    "label": "Headers",
    "schema": Object {
      "components": Object {
        "schemas": Object {
          "ProductStock": Object {
            "properties": Object {
              "test_param": Object {
                "items": Object {
                  "$ref": "#/components/schemas/SalesLine",
                },
                "type": "array",
              },
            },
            "type": "object",
            "x-readme-ref-name": "ProductStock",
          },
          "SalesLine": Object {
            "properties": Object {
              "stock": Object {
                "$ref": "#/components/schemas/ProductStock",
              },
            },
            "type": "object",
            "x-readme-ref-name": "SalesLine",
          },
          "dateTime": Object {
            "format": "date-time",
            "type": "string",
            "x-readme-ref-name": "dateTime",
          },
          "offset": Object {
            "properties": Object {
              "id": Object {
                "type": "string",
              },
              "rules": Object {
                "$ref": "#/components/schemas/rules",
              },
            },
            "type": "object",
            "x-readme-ref-name": "offset",
          },
          "offsetTransition": Object {
            "properties": Object {
              "dateTime": Object {
                "format": "date-time",
                "type": "string",
              },
              "offsetAfter": Object {
                "$ref": "#/components/schemas/offset",
              },
              "offsetBefore": Object {
                "$ref": "#/components/schemas/offset",
              },
            },
            "type": "object",
            "x-readme-ref-name": "offsetTransition",
          },
          "rules": Object {
            "properties": Object {
              "transitions": Object {
                "items": Object {
                  "$ref": "#/components/schemas/offsetTransition",
                },
                "type": "array",
              },
            },
            "type": "object",
            "x-readme-ref-name": "rules",
          },
        },
      },
      "properties": Object {
        "content": Object {
          "items": Object {
            "$ref": "#/components/schemas/SalesLine",
          },
          "type": "array",
        },
      },
      "required": Array [],
      "type": "object",
    },
    "type": "header",
  },
]
`;

exports[`parameters should convert parameters to JSON schema 1`] = `
Array [
  Object {
    "label": "Path Params",
    "schema": Object {
      "properties": Object {
        "petId": Object {
          "description": "Pet id to delete",
          "format": "int64",
          "type": "integer",
        },
      },
      "required": Array [
        "petId",
      ],
      "type": "object",
    },
    "type": "path",
  },
  Object {
    "label": "Headers",
    "schema": Object {
      "properties": Object {
        "api_key": Object {
          "type": "string",
        },
      },
      "required": Array [],
      "type": "object",
    },
    "type": "header",
  },
]
`;

exports[`polymorphism / discriminators should retain discriminator \`mapping\` refs when present 1`] = `
Array [
  Object {
    "label": "Body Params",
    "schema": Object {
      "components": Object {
        "schemas": Object {
          "OptionOneNoDisc": Object {
            "properties": Object {
              "discrim": Object {
                "type": "string",
              },
              "optionone": Object {
                "type": "number",
              },
            },
            "required": Array [
              "discrim",
            ],
            "type": "object",
            "x-readme-ref-name": "OptionOneNoDisc",
          },
          "OptionTwoNoDisc": Object {
            "properties": Object {
              "discrim": Object {
                "type": "string",
              },
              "optiontwo": Object {
                "type": "string",
              },
            },
            "required": Array [
              "discrim",
            ],
            "type": "object",
            "x-readme-ref-name": "OptionTwoNoDisc",
          },
        },
      },
      "discriminator": Object {
        "mapping": Object {
          "Option One": "#/components/schemas/OptionOneNoDisc",
          "Option Two": "#/components/schemas/OptionTwoNoDisc",
        },
        "propertyName": "discrim",
      },
      "oneOf": Array [
        Object {
          "properties": Object {
            "discrim": Object {
              "type": "string",
            },
            "optionone": Object {
              "type": "number",
            },
          },
          "required": Array [
            "discrim",
          ],
          "type": "object",
          "x-readme-ref-name": "OptionOneNoDisc",
        },
        Object {
          "properties": Object {
            "discrim": Object {
              "type": "string",
            },
            "optiontwo": Object {
              "type": "string",
            },
          },
          "required": Array [
            "discrim",
          ],
          "type": "object",
          "x-readme-ref-name": "OptionTwoNoDisc",
        },
      ],
    },
    "type": "body",
  },
]
`;

exports[`request bodies should convert request bodies to JSON schema (application/json) 1`] = `
Array [
  Object {
    "label": "Body Params",
    "schema": Object {
      "properties": Object {
        "category": Object {
          "properties": Object {
            "id": Object {
              "format": "int64",
              "type": "integer",
            },
            "name": Object {
              "type": "string",
            },
          },
          "type": "object",
          "x-readme-ref-name": "Category",
        },
        "id": Object {
          "format": "int64",
          "readOnly": true,
          "type": "integer",
        },
        "name": Object {
          "examples": Array [
            "doggie",
          ],
          "type": "string",
        },
        "photoUrls": Object {
          "items": Object {
            "type": "string",
          },
          "type": "array",
        },
        "status": Object {
          "description": "pet status in the store",
          "enum": Array [
            "available",
            "pending",
            "sold",
          ],
          "type": "string",
        },
        "tags": Object {
          "items": Object {
            "properties": Object {
              "id": Object {
                "format": "int64",
                "type": "integer",
              },
              "name": Object {
                "type": "string",
              },
            },
            "type": "object",
            "x-readme-ref-name": "Tag",
          },
          "type": "array",
        },
      },
      "required": Array [
        "name",
        "photoUrls",
      ],
      "type": "object",
      "x-readme-ref-name": "Pet",
    },
    "type": "body",
  },
]
`;

exports[`request bodies should convert request bodies to JSON schema (application/x-www-form-urlencoded) 1`] = `
Array [
  Object {
    "label": "Path Params",
    "schema": Object {
      "properties": Object {
        "petId": Object {
          "description": "ID of pet that needs to be updated",
          "format": "int64",
          "type": "integer",
        },
      },
      "required": Array [
        "petId",
      ],
      "type": "object",
    },
    "type": "path",
  },
  Object {
    "label": "Form Data",
    "schema": Object {
      "properties": Object {
        "name": Object {
          "description": "Updated name of the pet",
          "type": "string",
        },
        "status": Object {
          "description": "Updated status of the pet",
          "type": "string",
        },
      },
      "type": "object",
    },
    "type": "formData",
  },
]
`;

exports[`type sorting should return with a json schema for each parameter type (body instead of formData) 1`] = `
Array [
  Object {
    "label": "Path Params",
    "schema": Object {
      "properties": Object {
        "path parameter": Object {
          "type": "string",
        },
      },
      "required": Array [],
      "type": "object",
    },
    "type": "path",
  },
  Object {
    "label": "Query Params",
    "schema": Object {
      "properties": Object {
        "query parameter": Object {
          "type": "string",
        },
      },
      "required": Array [],
      "type": "object",
    },
    "type": "query",
  },
  Object {
    "label": "Body Params",
    "schema": Object {
      "properties": Object {
        "a": Object {
          "type": "string",
        },
      },
      "type": "object",
    },
    "type": "body",
  },
  Object {
    "label": "Cookie Params",
    "schema": Object {
      "properties": Object {
        "cookie parameter": Object {
          "type": "string",
        },
      },
      "required": Array [],
      "type": "object",
    },
    "type": "cookie",
  },
  Object {
    "label": "Headers",
    "schema": Object {
      "properties": Object {
        "header parameter": Object {
          "type": "string",
        },
      },
      "required": Array [],
      "type": "object",
    },
    "type": "header",
  },
]
`;

exports[`type sorting should return with a json schema for each parameter type (formData instead of body) 1`] = `
Array [
  Object {
    "label": "Path Params",
    "schema": Object {
      "properties": Object {
        "path parameter": Object {
          "type": "string",
        },
      },
      "required": Array [],
      "type": "object",
    },
    "type": "path",
  },
  Object {
    "label": "Query Params",
    "schema": Object {
      "properties": Object {
        "query parameter": Object {
          "type": "string",
        },
      },
      "required": Array [],
      "type": "object",
    },
    "type": "query",
  },
  Object {
    "label": "Cookie Params",
    "schema": Object {
      "properties": Object {
        "cookie parameter": Object {
          "type": "string",
        },
      },
      "required": Array [],
      "type": "object",
    },
    "type": "cookie",
  },
  Object {
    "label": "Form Data",
    "schema": Object {
      "properties": Object {
        "a": Object {
          "type": "string",
        },
      },
      "type": "object",
    },
    "type": "formData",
  },
  Object {
    "label": "Headers",
    "schema": Object {
      "properties": Object {
        "header parameter": Object {
          "type": "string",
        },
      },
      "required": Array [],
      "type": "object",
    },
    "type": "header",
  },
]
`;
